### DISCUSSION TITLE
How to encode measured EEG data to spike data

### DISCUSSION THREAD

**Discussion by hanbuck30** [jonescompneurolab/hnn-core#787] [Q&A]
![image](https://github.com/jonescompneurolab/hnn-core/assets/59501945/e4f92daf-1094-476d-b795-96a30e0f89ea)

![image](https://github.com/jonescompneurolab/hnn-core/assets/59501945/57502b35-f636-407e-87a9-dfc7f36f9ede)
I ran sample for the first time, and I get a file not found error. Even if I go into the mod folder, I don't have a .so or .dll file, where can I get it?

**Comment by jasmainak** [comment]
can you please provide more information how you did the installation and what kind of computer you have?

**Comment by hanbuck30** [comment]
I used the "git clone https://github.com/jonescompneurolab/hnn-core.git " command. And I'm using window 11, and I ran it on Vs code. Python version is 3.8 version and neurons are 8.2.
![image](https://github.com/jonescompneurolab/hnn-core/assets/59501945/e84b61c3-44f1-4fa3-8dec-7adf9adb3b68)

When I check the error, I check the error that requires .so or .dll in the mod folder, but when I check the git clone, I can't check the file in that format in mod

**Comment by jasmainak** [comment]
the mod files are built during installation using `nrniv` which is distributed using Neuron. You have to install neuron *before* installing HNN ... otherwise the mod files won't be built.

**Comment by hanbuck30** [comment]
![image](https://github.com/jonescompneurolab/hnn-core/assets/59501945/2862709f-b62b-47ca-afcd-68a6b5288df3)
I installed neuron before install HNN_core. but it still cause an same error

**Comment by rythorpe** [comment]
I'm assuming you followed the [instructions for an editable install](https://github.com/jonescompneurolab/hnn-core/blob/master/doc/contributing.rst#installing-editable-hnn-core)?

**Comment by hanbuck30** [comment]
Oh,  It works.  I think I forgot.
Thank you for your help.
One more question. I want to extract a code that encodes EEG data into a spike, which file should I refer to?

**Comment by rythorpe** [comment]
I'd recommend checking out our examples page [here](https://jonescompneurolab.github.io/hnn-core/stable/auto_examples/index.html). There are a few examples where we simulate a current dipole (underlying an EEG sensor signal) and then examine the spiking activity that emerged within the network.

**Comment by rythorpe** [comment]
I'm going to close this issue, but feel free to post on the repository's Discussions forum with any more questions that arise.

**Comment by hanbuck30** [comment]
I'm going to create spikes through the measured EEG data and use them for SNN, but I saw From MEG sensor-space data to HNN simulation.ipynb while looking at the samples you presented. Can I refer to this? I'd like you to tell me what to do with precautions.
Thank you.

**Comment by hanbuck30** [comment]
Thank you for answer @rythorpe
Hello, I referred to sample somato to use HNN_core through EEG. I checked that MRI data is needed for source localization. I tried to find the initial dipole and find the evoked drive through it as optimization, but can't I use HNN_core only with EEG data?

**Reply by rythorpe** [reply]
Perhaps I don't fully understand your question, but you can use hnn-core to interpret any EEG or MEG inverse solution provided it is in units of nano-Ampere-meters (nAm).

**Reply by hanbuck30** [reply]
I want to use hnn_core to convert the measured EEG into spikes. I have thought about the method by looking at the examples. I referred to the **plot_simulate_somato** and **optimize_evoked** files.

_I have written down the process and problems I have thought of._
1. Load the EEG and perform preprocessing.
2. To convert it into dipoles, it seems necessary to use the **apply_inverse** and **mne.read_labels_from_annot** functions. However, it looks like it is not possible to use only the EEG data obtained from the measuring equipment. What data is necessary for source reconstruction? Is it possible with just the montage?
3. The data attempted in **optimize_evoked** was with 1 trial (tmin=-.2, tmax=.17), but the optimization took 148 minutes. Generally, it is expected to take a very long time with measured EEG (Imagined EEG). Is there a way to reduce this time?

**Reply by rythorpe** [reply]
First off, it's important to note that inferring spiking activity from M/EEG signals is a highly ill-posed optimization problem. As such, HNN-core isn't designed to make a black-box EEG->spikes conversion, but rather to test specific mechanistic hypotheses on how M/EEG signals are generated. Much care should be taken when interpreting the results of our optimization routines.

Re: your questions:

2. To calculate the inverse solution you need the structural MRI for each participant. [MNE-Python's website](https://mne.tools/stable/index.html) is a great resource for understanding this process.
3. Optimization of HNN-core parameters intrinsically takes a long time because running a single simulation is so costly. If you're not doing this already, I encourage you to use the [`MPIBackend`](https://jonescompneurolab.github.io/hnn-core/stable/parallel.html#mpi) to parallelize your simulations and speed things up a little. Note that the optimization will still take a long time, but hopefully less of a long time :laughing:
